backend/search/models.py

```diff
@@ -10,18 +23,63 @@ class ComponentSearchResult(models.Model):
     last_accessed = models.DateField(null=True, editable=False)
     results = models.TextField(default='', editable=False)
     number_of_results = models.PositiveIntegerField(null=True, editable=False)
+    errors = models.TextField(default='', editable=False)
     completed_part = models.PositiveIntegerField(
         null=True, editable=False,
         help_text='Total size in KiB of databases for which the search has '
                   'been completed'
     )
 
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(fields=['xpath', 'component'],
+                                    name='componentsearchresult_uniqueness')
+        ]
+
     def __str__(self):
         return '"{}…" for {}'.format(self.xpath[:10], self.component)
 
     def perform_search(self):
         """Perform full component search and regularly update on progress"""
-        print('Non-implemented method perform_search() called')
+        if not check_xpath(self.xpath):
+            raise SearchError('Malformed XPath')
+        if not self.component_id:
+            raise SearchError('Field component not filled in')
+        databases_with_size = self.component.get_databases()
+        self.results = ''
+        self.completed_part = 0
+        self.number_of_results = 0
+        start_time = timer()
+        next_save_time = start_time + 1
+        matches = []
+        for database in databases_with_size:
+            size = databases_with_size[database]
+            query = generate_xquery_search(database, self.xpath)
+            try:
+                result = basex.perform_query(query)
+            except (OSError, UnicodeDecodeError) as err:
+                self.errors += 'Error searching database {}: ' \
+                    .format(database) + str(err) + '\n'
+                result = ''  # No break because completed_part is to be updated
+            try:
+                matches.extend(parse_search_result(
+                    result, self.component.slug, database)
+                )
+            except ValueError as err:
+                self.errors += 'Error parsing search result in database {}: ' \
+                    '{}\n'.format(database, err)
+            self.number_of_results = len(matches)
+            self.results = json.dumps(matches)
```

Comment:
This will be problematic for large result sets. Multiple copies of the dataset will be saved. This will also search everything even when a search is limited/aborted. It could be nice though, to be able to have some form of copy of the results available in the context of the MWE. Because the result sets are expected to be much smaller then, an approach such as this (or similar) would probably be viable.

Reply:
I also think that this may become problematic; it seems to work correctly but the database quickly becomes very large (e.g. 30 MiB for one ComponentSearchResult). One solution is, as Ben proposed, to only save the sentence IDs so that the results can quickly be obtained. Indeed it is also necessary that a search can be aborted.

I think that saving results will be necessary, though. Some treebanks (such as SoNaR) are simply too large to be searched in a short time. It would be ideal if the user could first get results from a small subset, so that they can test their query, and then get the opportunity to perform the full search. That search will then be performed in the background and the user can come back as soon as the search is finished. For semi-large treebanks, like EUROPARL, the current setup seems to work quite okay.

The reason that it searches and saves everything even if only the first 500 results are needed initially is because only counting is not much quicker than the full search. Also, these results can then be re-used in the analysis phase.

I think it will be good if we have a discussion about the design of the search process.

Reply:
It might be over-complicating things, but we could consider storing results in an in-memory cache (redis/memcached) with an LRU eviction policy.